### PR TITLE
Add implementation of `Select()`

### PR DESCRIPTION
**Pull request by djcarter85**
PR for https://github.com/ociaw/RandN/issues/11.

I decided to split this up even further, and just add support for `Select()` in this PR.

I've added a `MockDistribution`. This is quite similar to the `Singleton` that I added in the previous PR, but it allows you to control the behaviour of `Sample` and `TrySample`. I think this is a sensible way to test the behaviour of `Select` because we only care about how it maps the values from the underlying distribution.

There's a test with the comment `If sampling fails, the result from the input distribution should be ignored.`. This seems like the most sensible behaviour in this situation, because I imagine most underlying distributions will return `null` in the failure case and so the `selector` function is possibly not suitable to be applied. However, returning `default` in that case is giving me a compile warning of a possible null reference assignment. What do you think is the best way to deal with this? Would it be true to say that if `TrySample` returns `false` that the consumer shouldn't be reading the value of `result`?

### CODE REVIEW

**Comment by ociaw** [comment]
I've added a `MaybeNullWhen(false)` annotation to the out parameter of `TrySample` in 818814e3f051ee8fc0dd2af7575f6574b64b51df - can you see if the compiler still complains?

**Comment by djcarter85** [comment]
> I've added a `MaybeNullWhen(false)` annotation to the out parameter of `TrySample` in [818814e](https://github.com/ociaw/RandN/commit/818814e3f051ee8fc0dd2af7575f6574b64b51df) - can you see if the compiler still complains?

Merging `master` into my branch, I still get the compiler warning.

I then thought maybe I need to add the `MaybeNullWhen(false)` attribute to the out parameter in my implementation. But that gives me the compile error:

```
error CS0433: The type 'MaybeNullWhenAttribute' exists in both 'RandN.Core, Version=0.3.0.0, Culture=neutral, PublicKeyToken=null' and 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
```

Any idea what the right thing to do is?

**Comment by ociaw** [comment]
I think it should be working now - `MaybeNullWhenAttribute` isn't available in .NET Standard 2.0, so we kind of have to hack in support for it. I make the attribute `internal` for the `Core` assembly and copied it into the `RandN` assembly, so it should be working now.

**Comment by djcarter85** [comment]
That fixed it! I needed to add the attribute to my implementation to remove the compiler warning but that seems sensible anyway.

BTW I think the addition you made to the doc comment about `result` being unspecified is a great idea, as it makes it clear to the user that they should ignore the value when the return value is `false`.

File: src/RandN/Extensions/DistributionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        /// <typeparam name="TSource">The generic type of the input distribution.</typeparam>
+        /// <typeparam name="TResult">The generic type of the output distribution.</typeparam>
+        /// <param name="distribution">The distribution to be transformed.</param>
+        /// <param name="selector">The projection to be applied to values from the distribution.</param>
+        public static IDistribution<TResult> Select<TSource, TResult>(
+            this IDistribution<TSource> distribution,
+            Func<TSource, TResult> selector) =>
+            new SelectDistribution<TSource, TResult>(distribution, selector);
+
+        private class SelectDistribution<TSource, TResult> : IDistribution<TResult>
```
Comment (ociaw) [inline src/RandN/Extensions/DistributionExtensions.cs:32]: I try to make every class `sealed` unless I explicitly need inheritance for some reason.
(Also testing out GitHub suggestions?)
```suggestion
        private sealed class SelectDistribution<TSource, TResult> : IDistribution<TResult>
```

File: src/RandN/Extensions/DistributionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        /// IDistribution&lt;int&gt; evenNumbersDistribution =
+        ///     from i in Uniform.NewInclusive(0, 5)
+        ///     select i * 2;
+        /// </code>
+        /// </example>
+        /// <typeparam name="TSource">The generic type of the input distribution.</typeparam>
+        /// <typeparam name="TResult">The generic type of the output distribution.</typeparam>
+        /// <param name="distribution">The distribution to be transformed.</param>
+        /// <param name="selector">The projection to be applied to values from the distribution.</param>
+        public static IDistribution<TResult> Select<TSource, TResult>(
```
Comment (ociaw) [inline src/RandN/Extensions/DistributionExtensions.cs:27]: Is there any reason this can't return the `SelectDistribution` type outright? There are potential (though minor) performance benefits since methods calls won't need to be a virtual. The caller could always cast it to an `IDistribution<TResult>` later if they needed to.

File: src/RandN/Extensions/DistributionExtensions.cs (hunk 1 above)
Comment (djcarter85) [reply src/RandN/Extensions/DistributionExtensions.cs]: Good call - I've applied the suggestion.

File: src/RandN/Extensions/DistributionExtensions.cs (hunk 2 above)
Comment (djcarter85) [reply src/RandN/Extensions/DistributionExtensions.cs:27]: Currently the only reason why this is needed is that the `SelectDistribution` type is private, but I could easily make it public (maybe with an internal constructor).

One benefit of returning the interface rather than the concrete type is that the caller probably only cares about getting an object which implements the interface. I think that's why the standard LINQ `Select()` returns an `IEnumerable<TResult>`. I don't know how much of a big deal that is though.

The benefit of returning the concrete type is, as you say, improved performance. There's also the fact that most other methods in this library return the concrete type, so consistency with that would be good.

TL;DR I'm happy to return the concrete type if you think it's best!

Comment (djcarter85) [reply src/RandN/Extensions/DistributionExtensions.cs:27]: Just tried it out locally. One oddity is that the public interface of `SelectDistribution<TSource, TResult>` only cares about one of the type parameters; namely `TResult`. Not sure if that sways things or not.

Comment (ociaw) [reply src/RandN/Extensions/DistributionExtensions.cs:27]: Those are good points - I was thinking of how ImmutableArray has its own extension methods that are optimized for it instead of the standard LINQ expressions. But I don't think that's relevant here since this would be the generalized method. So I'm leaning back towards returning IDistribution<TResult> again.
